### ISSUE TITLE
balance wrong after unloading wallet and loading it again

### ISSUE THREAD

**Issue by benwilliams** [bitcoin/bitcoin#15591]
Have two wallets used by this instance of bitcoind. The default wallet has balance 0.02861762.
Initially reported balance is correct but after unloading the wallet and then loading it again, the balance is wrong. Neither of the wallets have the reported balance.

I expected the balance to be the same as initially reported.
```
btc@ubuntu:~$ bitcoin-cli  -getinfo
{
  "version": 170100,
  "protocolversion": 70015,
  "walletversion": 159900,
  "balance": 0.02861762,
...

btc@ubuntu:~$ bitcoin-cli  loadwallet wallet2
btc@ubuntu:~$ bitcoin-cli unloadwallet ""
btc@ubuntu:~$ bitcoin-cli loadwallet ""
btc@ubuntu:~$ bitcoin-cli unloadwallet wallet2
btc@ubuntu:~$ bitcoin-cli  -getinfo
{
  "version": 170100,
  "protocolversion": 70015,
  "walletversion": 159900,
  "balance": 0.01012607,
...
```
The balance 0.01012607 is wrong and I have no idea where that number came from.
Restarting the process results in correct balance again.

ubuntu package for bionic

**Comment by benwilliams** [comment]
The issue is no longer happening. I repeated the exact same steps as before and it is working correctly now and unable to replicate it again. I still have the terminal window open to prove it occurred and I'm not going crazy.
The wallets were originally on Intel hardware and were moved onto a newer OS on AMD hardware. Surely that wouldn't have caused issues?

**Comment by MeshCollider** [comment]
I've managed to replicate similar inconsistency following your method:

```
$ ./bitcoind -regtest -datadir=../datadir &
$ ./bitcoin-cli -regtest -datadir=../datadir -getinfo
{
  "version": 189900,
  "protocolversion": 70015,
  "blocks": 200,
  "chain": "regtest",
  "walletversion": 169900,
  "balance": 4999.99996680,
   ...
}
$ ./bitcoin-cli -regtest -datadir=../datadir loadwallet test
$ ./bitcoin-cli -regtest -datadir=../datadir unloadwallet ""
$ ./bitcoin-cli -regtest -datadir=../datadir loadwallet ""
$ ./bitcoin-cli -regtest -datadir=../datadir unloadwallet test
$ ./bitcoin-cli -regtest -datadir=../datadir -getinfo
{
  "version": 189900,
  "protocolversion": 70015,
  "blocks": 200,
  "chain": "regtest",
  "walletversion": 169900,
  "balance": 4950.00000000,
   ...
}
```

Notice the second balance is nearly 50 BTC less than the first. I'll look into this.

**Comment by MarcoFalke** [comment]
Assigned milestone 0.18.0

**Comment by MarcoFalke** [comment]
Is this a regression from 0.17.1?

**Comment by promag** [comment]
@MarcoFalke the issue exists since `loadwallet` was added, so should be fixed since 0.17.2 and 0.18 rc3.

The problem is that dynamically loaded wallet don't see mempool transactions.

It's more clear if you use `getwalletinfo` (instead of `-getinfo`) and observe `unconfirmed_balance`.

Not sure what should be the right fix. There is
https://github.com/bitcoin/bitcoin/blob/68520597ccf8ff3f6e8a7ad6869b06bf2012ae8a/src/wallet/wallet.cpp#L4399-L4404
but should it also do?
```
     // Add mempool transactions that aren't yet in the wallet
```

**Comment by promag** [comment]
To reproduce, use one wallet `test1` to send a transaction to another wallet `test2` while `test2` is unloaded, then load `test2`:

```
# generate on test1
bitcoind -regtest -wallet=test1 &
bitcoin-cli -regtest getnewaddress
bitcoin-cli -regtest generatetoaddress 101 $(bitcoin-cli -regtest getnewaddress)

# create test2, get receiving address and unload the wallet
bitcoin-cli -regtest createwallet test2
A=$(bitcoin-cli -regtest -rpcwallet=test2 getnewaddress)
bitcoin-cli -regtest -rpcwallet=test2 unloadwallet

# send value to the above address and check balance
bitcoin-cli -regtest sendtoaddress $A 1
bitcoin-cli -regtest loadwallet test2
bitcoin-cli -regtest -rpcwallet=test2 getwalletinfo
{
  "walletname": "test2",
  "walletversion": 169900,
  "balance": 0.00000000,
  "unconfirmed_balance": 0.00000000,
  ...

# restart now with wallet test2 and check balance
bitcoin-cli -regtest stop
bitcoind -regtest -wallet=test2
bitcoin-cli -regtest getwalletinfo
{
  "walletname": "test2",
  "walletversion": 169900,
  "balance": 0.00000000,
  "unconfirmed_balance": 1.00000000,
```
